doc/rst/technotes/dsi.rst

```diff
@@ -954,5 +954,185 @@ implicitly for each ``GlobalDomain`` object and so will be local in any case.
+
+.. method:: proc GlobalArray.doiBulkTransferFromKnown(myDom:domain, otherClass, otherDom:domain) : bool
+
+  Perform an array assignment from ``otherClass`` to ``this``
+
+.. method:: proc GlobalArray.doiBulkTransferToAny(myDom:domain, otherClass, otherDom:domain) : bool
+
+  Perform an array assignment from ``this`` to ``otherClass``.
+
+.. method:: proc GlobalArray.doiBulkTransferFromAny(myDom:domain, otherClass, otherDom:domain) : bool
+
+  Perform an array assignment from ``otherClass`` to ``this``
+
+Each method shares these arguments:
+
+* ``myDom:domain`` - the indices to be transferred to/from ``this``
+
+* ``otherClass`` - a class inheriting from ``BaseArr`` that represents the
+  other array in the assignment
+
+* ``otherDom:domain`` - the indices to be transferred to/from ``otherClass``
+
+Callers of these methods must satisfy the following guarantees for these
+arguments:
+
+* ``this.rank == myDom.rank``
+
+* ``otherClass.rank == otherDom.rank``
+
+* ``myDom.size == otherDom.size``
+
+.. note::
+
+  ``myDom`` and ``otherDom`` are not required to be distributed, even if the
+  corresponding arrays are distributed. They simply represent the indices
+  involved in the transfer.
+
+Each method returns a ``bool`` indicating whether the assignment occurred. This
+allows domain maps to resort to the Chapel implementation of array assignment
+if they are incapable or unwilling to perform the assignment themselves.
+
+Known and Any methods
+---------------------
+
+There are two kinds of transfers accounted for in this interface.
+
+The first kind of transfer is a ``Known`` transfer. These are transfers that
+the implementer believes to be optimal. Typically this means that the
+implementer knows the type and internals of the other domain map involved in
+the transfer.
+
+The second kind of transfer is an ``Any`` transfer. These are transfers to/from
+domain maps whose type and internals are **not** known. These transfers rely on
+a common interface, which may involve unnecessary overhead compared to a
+``Known`` transfer. For example, it is likely that many domain maps will
+implement ``Known`` transfers to/from Chapel's default rectangular arrays. It
+is also likely that many domain maps will be implemented using default
+rectangular arrays. Knowing this, implementers of ``Any`` transfers can
+implement in terms of ``Known`` transfers to/from their own default rectangular
+arrays as a common interface.
+
+This interface recognizes these kinds of transfers because otherwise it is
+difficult to determine which distribution implements the best transfer.
+
+For example, the author of ``PopularDist`` could implement ``Any`` methods for
+better performance with unknown third-party domain maps. The author of
+``NewDist`` could implement ``Known`` methods to ``PopularDist`` due to its
+popularity. By doing so the author of ``NewDist`` has informed Chapel that it
+knows better than ``PopularDist`` and a more optimal transfer occurs.
```

Comment:
Comments from the peanut gallery:

Given ``A = B``, is it defined whether the call is to ``A.transferFrom(B)`` or ``B.transferTo(A)``?
Or, how are the pair of directions meant to be used?  Having both suggests that either there are cases the compiler will generate either; or it's to allow a ``this.transferFrom(that)`` to turn around and call ``that.transferTo(this)``.  But the latter seems ripe with opportunity for unintentional recursion without guidance to implementers.  (e.g., "``From``s can call ``To``s, but not vice versa.")  Or is it only that the compiler will always try ``A.from(B)`` and if that fails, ``B.to(A)``?

Reply:
The order is currently:
1) Try either ``Dest.transferFromKnown(Src)`` or ``Src.transferToKnown(Dest)``
2) Try either ``Dest.transferFromAny(Src)`` or ``Src.transferToAny(Dest)``

We currently do not define whether ``To`` or ``From`` is attempted first. The compiler will not generate any of these methods. Our internal implementation of arrays will simply attempt to resolve and then call these methods.

You make a good point about possible recursion, I'll add a warning calling out this problematic pattern.

I'll also add a note about undefined To/From preference.